### DISCUSSION TITLE
How to create custom package repo

### DISCUSSION THREAD

**Discussion by oraqlle** [vector-of-bool/dds#103] [Q&A] [answered]
Is it possible to build a repository locally of different packages and libraries and then add the repo (with the repo.db and the pkg/ directory) to GitHub and have DDS pull the information from GitHub and use it to resolve information. If so how would this be done. 

I also reckon the documentation should include these steps to creating your own database repository, if using GitHub (or other remote repositories) is possible. I'll create a formal request in the Issues Page. 

*\-* Regards.

**Comment by vector-of-bool** [comment, marked as answer]
Hello!

Sorry for the delay.

Repositories are directories that contain only static files and can be served by any basic HTTP server, so it is entirely possible to serve one from GitHub. In the alpha.6 release of `dds`, repositories are created and managed by the `dds repoman` subcommand, but it is very clunky and only the basics I needed to get the initial public repository available.

However, in the upcoming beta release (which has been a _long time_ coming), I'm doing significant changes to the way packages and repositories are handled. It will still be just a static directory of files, but the subcommands, usage, and shape of the `repo.db` are all changing up. I will probably not have the documentation ready immediately with the beta release, as repository management is a more advanced feature, but I do intend to fully document the feature.

---

Here's a very very very basic how-to on the existing (soon to be replaced) `dds repoman`:

1. Use `dds repoman init <repo-dir>` to initialize a new repository.
2. Use `dds pkg create` on a dds-buildable directory to create a `.tar.gz` source distribution archive of that package.
3. Use `dds repoman import <repo-dir> /path/to/generated-pkg.tar.gz` to import the package into the repository.
4. Serve `<repo-dir>` from an HTTP server, and use `dds pkg repo add` to add that directory as a repository.

Beware that all of the above will change in the upcoming beta release, but you should be able to experiment with `dds` repositories using this process.

**Reply by oraqlle** [reply]
Thank you very much for your response. I actually got this to work somewhat is an existing static site I have running. I only ask not because the documentation is unclear but I'm hopeless in the web development space. It's okay it clanky for now, I'm really looking forward to what's to come and DDS is already super powerful (and simple to work with). Keep up the awesome work. 

And I'd be happy to help out and contribute to this if you need. I don't have much experience but I thought I'd offer because I want to see this idea fulfilled.

**Comment by david-bakin** [comment]
Suppose you want to do this on a local file system; will this work without a local web server by using a `file:` url?

**Reply by vector-of-bool** [reply]
In the alpha.6 release, file:// URLs are not supported. In the upcoming beta release filepaths and file:// URLs will be fully supported for repositories.
